core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -70,23 +72,174 @@ object H2OSchemaUtils {
     StructType(types)
   }
 
-  def flattenSchema(schema: StructType, prefix: String = null, nullable: Boolean = false): StructType = {
+  def flattenDataFrame(df: DataFrame): DataFrame = {
+    val schema = flattenSchema(df)
+    flattenDataFrame(df, schema)
+  }
+
+  def flattenDataFrame(df: DataFrame, flatSchema: StructType): DataFrame = {
+    implicit val rowEncoder = RowEncoder(flatSchema)
+    val numberOfColumns = flatSchema.fields.length
+    val flatSchemaIndexes = flatSchema.fields.map(_.name).zipWithIndex.toMap
+    val originalSchema = df.schema
+    df.map[Row]{ r: Row =>
+      val result = ArrayBuffer.fill[Any](numberOfColumns)(null)
+      val fillMethod = fillBuffer(flatSchemaIndexes, result, r)(_, _, _)
+      originalSchema.fields.zipWithIndex.foreach{ case (f, i) => fillMethod(f, i, None)}
+      Row.fromSeq(result)
+    }
+  }
+
+  private def fillBuffer
+      (flatSchemaIndexes: Map[String, Int], buffer: ArrayBuffer[Any], row: Row)
+      (field: StructField, index: Int, prefix: Option[String] = None): Unit = {
+    val StructField(name, dataType, _, _) = field
+    val qualifiedName = getQualifiedName(prefix, name)
+    if(!row.isNullAt(index)) {
+      dataType match {
+        case BinaryType =>
+          fillArray(flatSchemaIndexes, buffer, row, ByteType, index, qualifiedName)
+        case MapType(_, valueType, containsNull) =>
+          val map = row.getMap[Any, Any](index)
+          val subRow = Row.fromSeq(map.values.toSeq)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          map.keys.zipWithIndex.foreach{ case (k, i) =>
```

Comment:
NIT: Missing space after foreach.

Also what about renaming (k, i) to (key, idx) ?

I prefer the longer names for easier readability :D